src/aten/TensorAdvancedIndexing.cpp

```diff
@@ -73,4 +75,110 @@ Tensor& XPUNativeFunctions::masked_fill_(
   return XPUNativeFunctions::masked_fill_(self, mask, value.item());
 }
 
+void index_func_meta_impl(
+    Tensor& result,
+    const Tensor& self,
+    int64_t dim,
+    const Tensor& index,
+    const Tensor& source,
+    c10::string_view func) {
+  auto numel = index.numel();
+
+  TORCH_CHECK_INDEX(
+      index.dim() <= 1,
+      func,
+      "_(): Index is supposed to be a vector, but got dim: ",
+      index.dim(),
+      " with type: ",
+      index.scalar_type(),
+      " and size: ",
+      index.sizes());
+  TORCH_CHECK(
+      index.scalar_type() == ScalarType::Long ||
+          index.scalar_type() == ScalarType::Int,
+      func,
+      "_(): Expected dtype int32/int64 for index but got: ",
+      index.scalar_type());
+  TORCH_CHECK(
+      self.scalar_type() == source.scalar_type(),
+      func,
+      "_(): self (",
+      self.scalar_type(),
+      ") and source (",
+      source.scalar_type(),
+      ") must have the same scalar type");
+  TORCH_CHECK(
+      dim == 0 || dim < source.dim(),
+      func,
+      "_(): Indexing dim ",
+      dim,
+      " is out of bounds of the source tensor with dim ",
+      source.dim());
+  TORCH_CHECK(
+      numel == (source.dim() == 0 ? 1 : source.size(dim)),
+      func,
+      "_(): Number of indices (",
+      numel,
+      ") should be equal to source.size(dim): (",
+      source.size(dim),
+      "), for dim: ",
+      dim);
+
+  auto self_sizes = self.sizes().vec();
+  auto source_sizes = source.sizes().vec();
+  if (source.dim() != 0 && self.dim() != 0) {
+    self_sizes.erase(self_sizes.begin() + dim);
+    source_sizes.erase(source_sizes.begin() + dim);
+  }
+  TORCH_CHECK(
+      self_sizes == source_sizes,
+      "source tensor shape must match self tensor shape, excluding the specified dimension. Got self.shape = ",
+      self.sizes(),
+      " source.shape = ",
+      source.sizes());
+
+  bool is_defined = result.defined();
+
+  // set_output_raw_strided
+  auto options = self.options();
+  auto sizes = self.sizes();
+  at::xpu::resize_out(result, sizes, {}, options);
+  if (is_defined) {
+    at::assert_no_internal_overlap(result);
+    at::assert_no_overlap(result, index);
+    at::assert_no_overlap(result, source);
```

Comment:
Have all these operations been implemented for XPU?

Reply:
They are all host functions. No need for XPU implementation.